File: src/pathfinding.js
import { neighborsOf } from './hex.js';

// Step-time (ステップ) and fuel (燃料) cost per terrain, per the rules table.
export const TERRAIN_STEP_COST = {
  plain: 2,
  mountain: 3,
  road_smooth: 1,
  road_congested: 2,
  road_jammed: 4,
};

export const TERRAIN_FUEL_COST = {
  plain: 1,
  mountain: 2,
  road_smooth: 2,
  road_congested: 2,
  road_jammed: 2,
};

export function isWalkable(terrain) {
  return terrain !== 'pond';
}

// Resolves a cell's effective terrain key (folding road state into the
// terrain type) so cost tables can be looked up with a single key.
export function terrainKey(cellTerrain, roadState) {
  if (cellTerrain !== 'road') return cellTerrain;
  return `road_${roadState || 'smooth'}`;
}

export function stepCost(cellTerrain, roadState) {
  return TERRAIN_STEP_COST[terrainKey(cellTerrain, roadState)];
}

export function fuelCost(cellTerrain, roadState) {
  return TERRAIN_FUEL_COST[terrainKey(cellTerrain, roadState)];
}

// Dijkstra shortest path (by step-time) from `startId` to `targetId` over
// the walkable cells of the map. `roadStateOf(cellId)` resolves the current
// road congestion state for a cell. Returns an array of cell ids from
// start to target inclusive, or null if unreachable.
export function findPath(startId, targetId, { width, height, terrainOf, roadStateOf }) {
  if (startId === targetId) return [startId];

  const dist = new Map([[startId, 0]]);
  const prev = new Map();
  const visited = new Set();

  // Simple array-backed priority queue; grids here are small (<=1024 cells).
  const queue = [[0, startId]];

  while (queue.length) {
    queue.sort((a, b) => a[0] - b[0]);
    const [d, id] = queue.shift();
    if (visited.has(id)) continue;
    visited.add(id);
    if (id === targetId) break;

    // Step time (and fuel, elsewhere) to leave a cell is determined by the
    // terrain of the cell being departed, not the destination — per the
    // rules ("移動命令を受けた時点での地形に応じた燃料を消費します").
    const originTerrain = terrainOf(id);
    const cost = stepCost(originTerrain, roadStateOf ? roadStateOf(id) : 'smooth');
    for (const nb of neighborsOf(id, width, height)) {
      const nbTerrain = terrainOf(nb);
      if (!isWalkable(nbTerrain)) continue;
      const nd = d + cost;
      if (!dist.has(nb) || nd < dist.get(nb)) {
        dist.set(nb, nd);
        prev.set(nb, id);
        queue.push([nd, nb]);
      }
    }
  }

  if (!dist.has(targetId)) return null;

  const path = [targetId];
  let cur = targetId;
  while (cur !== startId) {
    cur = prev.get(cur);
    path.push(cur);
  }
  path.reverse();
  return path;
}

File: visualizer/pathfinding.py
"""地形コストを考慮した Dijkstra 経路探索。

ルール上、移動のステップ数・燃料は「移動命令を受けた時点での地形」
（=出発セルの地形）で決まるため、辺のコストは出発側セルに紐づく。
"""

import heapq

from .hexgrid import neighbors

# 地形キー → 移動ステップ数（表1）
STEP_COST = {
    "plain": 2,
    "mountain": 3,
    "road_smooth": 1,
    "road_congested": 2,
    "road_jammed": 4,
}

# 地形キー → 消費燃料（表1）
FUEL_COST = {
    "plain": 1,
    "mountain": 2,
    "road_smooth": 2,
    "road_congested": 2,
    "road_jammed": 2,
}

ROAD_STATES = ("smooth", "congested", "jammed")


def terrain_key(terrain: str, road_state: str = "smooth") -> str | None:
    """セルの地形と道路状態から STEP_COST/FUEL_COST のキーを返す。池は None。"""
    if terrain == "pond":
        return None
    if terrain == "road":
        return f"road_{road_state}"
    return terrain


def dijkstra(start: int, width: int, height: int, key_of) -> tuple[dict, dict]:
    """start から全セルへの最短ステップ数を返す。

    key_of(cell) はそのセルの地形キー（terrain_key の戻り値、池は None）。
    戻り値は (dist, prev)。到達不能セルは dist に含まれない。
    """
    dist = {start: 0}
    prev: dict[int, int] = {}
    heap = [(0, start)]
    done: set[int] = set()
    while heap:
        d, cell = heapq.heappop(heap)
        if cell in done:
            continue
        done.add(cell)
        key = key_of(cell)
        if key is None:
            continue
        leave_cost = STEP_COST[key]
        for nb in neighbors(cell, width, height):
            if key_of(nb) is None:
                continue
            nd = d + leave_cost
            if nd < dist.get(nb, 1 << 30):
                dist[nb] = nd
                prev[nb] = cell
                heapq.heappush(heap, (nd, nb))
    return dist, prev


def reconstruct_path(prev: dict, start: int, goal: int) -> list[int] | None:
    """dijkstra の prev から start→goal の経路（両端含む）を復元する。"""
    if goal == start:
        return [start]
    if goal not in prev:
        return None
    path = [goal]
    while path[-1] != start:
        path.append(prev[path[-1]])
    path.reverse()
    return path


def path_step_cost(path: list[int], key_of) -> int:
    return sum(STEP_COST[key_of(c)] for c in path[:-1])


def path_fuel_cost(path: list[int], key_of) -> int:
    return sum(FUEL_COST[key_of(c)] for c in path[:-1])
